### ISSUE TITLE
Smart lock integration for Alarmo

### ISSUE THREAD

**Issue by oded996** [nielsfaber/alarmo#724]
### Checklist
- [X ] I checked for similar existing requests (both open and closed) before posting.
- [ X] My request is generic, other users may benefit from it too.

### Proposal
The proposal is to add native Smart Lock support in Alarmo.
- In Alarmo UI, users should be able to select which locks should be managed by Alarmo. See proposed sketch UI ![mock](https://github.com/nielsfaber/alarmo/assets/58013633/5f533fd1-aff1-4b1a-a934-a9d5a96334b2)
- User codes will automatically be synced form Alarmo to selected smart locks.  Giving the benefit of managing user codes from a single place
- Add the option to disarm the alarm when user enters correct pin code into a lock.

This is similar to how Nest Secure and Ring alarm systems works with Zwave smar locks.

### Additional info
- Due to protocol limitation, not all smart locks can be managed by home assistant.  So we should start with ZWave (ZWave-JS).
- This could mean a braking change to Alarmo - Pin codes are currently stored hashed.  In order to sync pin codes to locks, it may need to store pin codes in plain-text instead
- Keymaster (home assistant addon) does offer the ability to manage pin codes for smart locks.  It is far from easy to use and not able to work well with Alarmo.  Integrating it to Alarmo UI will be much easier to use.
-

**Comment by nielsfaber** [comment]
Although I am not a user of smart locks myself, I can imagine that some users would like to have Alarmo interact with them.
So let's discuss ideas and see if there's anything that should be added to Alarmo.

> In Alarmo UI, users should be able to select which locks should be managed by Alarmo.

What do you mean exactly with 'managed by Alarmo'?
Alarmo is intended as a burglar alarm, not as a lock manager.
I would like to avoid adding functionality that is out of scope, so as previously mentioned I am at this point a hesitant to add a complete page dedicated to locks.

> User codes will automatically be synced form Alarmo to selected smart locks. Giving the benefit of managing user codes from a single place

I agree it would be nice if Alarmo can store and validate all codes, including those used for locks.

> This could mean a braking change to Alarmo - Pin codes are currently stored hashed. In order to sync pin codes to locks, it may need to store pin codes in plain-text instead

This is going to be a problem.
The pincodes should not be stored in a text file since this would compromise security.
I would expect any software to use hashing / encryption on any pincode or password to avoid leaking sensitive data.
I am not open to changing the current way passwords are stored within Alarmo.

We can use Alarmo for validating a code, but not have Alarmo send stored codes to other integrations/devices.
It could be an option to create a service like `alarmo.validate_code` which would give a result by sending an event in HA.
If this can be connected to existing lock integrations, I'm all up for it.

There is already something similar in place (in the form of a blueprint) for Zigbee / Z-wave keypads, see for example here: https://community.home-assistant.io/t/zigbee2mqtt-control-alarmo-via-keypad/420016.

> Due to protocol limitation, not all smart locks can be managed by home assistant. So we should start with ZWave (ZWave-JS).

I think we should not limit ourselves to only Z-Wave, but consider other platforms as well, at least Zigbee I'd say. 
As you can see [here](https://analytics.home-assistant.io/integrations/), only 10% of HA users use Z-Wave, against 40% with Zigbee2MQTT + 20% with ZHA.
We should focus on a function that would be beneficial for the majority of users.

Is there any documentation available on how to interact (e.g. validate pin codes) with a ZWave JS smart lock?

> Add the option to disarm the alarm when user enters correct pin code into a lock.

This can be a very useful feature, but can already be achieved via an automation.
I have ideas for adding a new section under 'actions'  for the automatic arming / disarming alarmo via remotes, tags, etc. 
Locks can definitely also benefit from this feature.
I would also consider automation in the opposite way:  operating a lock (i.e. locking a door) when the alarm is armed, which could be useful for something like a backdoor, which is not used for exiting.

In summary, as I understand we are considering 2 different functionalities:
1) Using Alarmo for validating pin codes entered on a smart lock
2) Automatic arming / disarming Alarmo when a lock is locked / unlocked (and perhaps in opposite direction as well, so have Alarmo trigger the locking/unlocking of a lock)

I think both can be useful additions, but I am concerned whether the first one is feasible by architectural limitations (the fact that stored codes are not accessible from outside of Alarmo).
I'd love to hear your thoughts on this.

**Comment by oded996** [comment]
Thank you [nielsfaber](https://github.com/nielsfaber).

These are all valid points. 
One option is to create a completely separate addon for managing smart locks and have it seamlessly interact with Alarmo using events. (e.g. disarm event)

The piece that is missing is that Alarmo, in addition to being a burglar alarm, it is also a user management tool.  This is where one could manage who can arm/disarm the alarm. 
The idea is to make it easier and have one place to manage users for both the alarm and smart locks.
In order to do that, the locks management tool needs access to the user/code list.

I can think of multiple options:
1) Build locks management into Alarmo 
- Due to the nature of syncing to locks, there need to be a way to load the raw codes 
2) Expose API from Alarmo so other HA components will be able to read the user list
- Not sure how to do it in a secure way as anything in HA will have access to user codes
3) Expose API form Alarmo so other HA components will be able to add/delete users.  This way we could create a separate addon to manage users and it will be optional for users who wants to integrate user lists with multiple addons
4) Extract user management into a separate addon or perhaps use HA people management tool in some way. 
- This is probably a lot of work.

I believe that options (1) and (3) are viable.  Let me know what you think

Regarding security of codes in storage, I am not sure what is the best way to deal with it. There are a few things to consider:
- If locks are synced to ZWave locks, any component in HA is able to use ZWave APIs to read the raw codes from locks.
- Storing hashed versions of the codes adds a layer of security, but due to the low entropy of codes (usually 4-6 digits), it is fairly easy brute force.
- One option could be to change the current hash (one-way function) to symmetric encryption. This will allow Alarmo to sync raw codes if needed.


Re Zwave locks:  I agree that it should support other protocols.  I am suggesting to start with ZWave because there are way more ZWave smart locks in the market and this protocol allows HA to manage codes.  Wifi/Bluetooth currently do not.
Zigbee is very popular for HA users, but there are not many locks that use Zigbee. We can add support as a second phase.


The high level goal is to bring HA in par with other alarm systems in the market that offers smart lock integrations
